spec/tictactoe_spec.rb

```diff
@@ -76,43 +76,105 @@ def is_row_outputed(io:, row:)
+        expect(updated_board[desired_position]).to eql player_token
       end
     end
 
     context 'when the user tries to place a token in an occupied space' do
       it "should prompt the user that it's taken" do
-        fakeIO = FakeIO.new( stdin: desired_position )
+        fakeIO = FakeIO.new(stdin: desired_position)
+        game = TicTacToe.new(io: fakeIO)
+
+        game.begin_player_turn
+
+        fakeIO.stdin = desired_position
+        game.begin_player_turn
+
+        user_is_prompted_position_taken = fakeIO.stdout.grep(/[P|p]osition.*taken/).any?
+        expect(user_is_prompted_position_taken).to eql true
+      end
+      it 'should prompt the user to choose another position' do
+        fakeIO = FakeIO.new(stdin: desired_position)
         game = TicTacToe.new(io: fakeIO)
 
         game.begin_player_turn
 
         fakeIO.stdin = desired_position
         game.begin_player_turn
 
-        user_is_prompted_position_taken = fakeIO.stdout.grep(/[P|p]osition.*taken/).any? 
-        expect( user_is_prompted_position_taken  ).to eql true
+        user_is_prompted_to_select_space = fakeIO.stdout.grep(/[S|s]elect another position/).any?
+        expect(user_is_prompted_to_select_space).to eql true
+      end
+    end
+
+    context 'when the user input is invalid' do
+      context 'if the input is outside of the specified range' do
+        it 'prompt the user to choose another position' do
+          fakeIO = FakeIO.new(stdin: -1)
+          game = TicTacToe.new(io: fakeIO)
+
+          game.begin_player_turn
+
+          user_is_prompted_position_invalid = fakeIO.stdout.grep(/[I|i]nvalid position/).any?
+          user_is_prompted_to_select_space  = fakeIO.stdout.grep(/[S|s]elect another position/).any?
+          expect(user_is_prompted_to_select_space).to eql true
+          expect(user_is_prompted_position_invalid).to eql true
+        end
+
+        it 'should not place the token on the board' do
+          fakeIO = FakeIO.new(stdin: -1)
+          game = TicTacToe.new(io: fakeIO)
+
+          updated_board = game.begin_player_turn
+
+          expect(updated_board.include?('X')).to_not eql true
+        end
+
+        it 'should not end the player\'s turn until a valid input is entered' do
+          fakeIO = FakeIO.new(stdin: -1)
+          game = TicTacToe.new(io: fakeIO)
+
+          player_one_token = game.current_player
+
+          game.begin_player_turn
+          expect(game.current_player).to eql player_one_token
+          fakeIO.stdin = 5
+          game.begin_player_turn
+
+          expect(game.current_player).to_not eql player_one_token
+        end
+
+        it 'should re-render the board' do
+          fakeIO = FakeIO.new(stdin: -1)
+          game = TicTacToe.new(io: fakeIO)
+
+          current_token = game.current_player
+          game.begin_player_turn
+          rendered_board = fakeIO.stdout[+(fakeIO.stdout.length - 5)..(fakeIO.stdout.length)]
```

Comment:
This line here is a bit difficult to follow. Why do you need to index the `stdout` in this way? Is there a more expressive way of parsing the output string?

Reply:
I'm using stdout to get the last board rendered. Each row is an element, there are 5 rows (counting the line dividers) in total. For a more expressive way, I can consolidate a board into a string and perform an include to see if the board being tested in included in `stdout`

Reply:
see my comment above about leveraging fakeIO to make this test easier to follow.

Reply:
I am going to resolve this convo since the above `rendered_board` line isn't there anymore. I'm not sure how to update the code that's in the convos